be/src/storage/rowset/plain_page.h

```diff
@@ -166,10 +166,11 @@ class PlainPageDecoder : public PageDecoder {
         // find the first value >= target. after loop,
         // - left == index of first value >= target when found
         // - left == _num_elems when not found (all values < target)
+        auto type_info = get_type_info(Type);
         while (left < right) {
             size_t mid = left + (right - left) / 2;
             mid_value = &_data[PLAIN_PAGE_HEADER_SIZE + mid * SIZE_OF_TYPE];
-            if (TypeTraits<Type>::cmp(mid_value, value) < 0) {
```

Comment:
It looks like a virtual function call will be generated here, and I'm not sure the impact

Reply:
concern +1

Reply:
I've disassemble the `starrocks::PlainPageDecoder<(starrocks::FieldType)1>::seek_at_or_after_value(void const*, bool*)` functions, which shows that all virtual function calls are eliminated and inlined. 

The call stack is:
```c++
PlainPageDecoder::seek_at_or_after_value() 
  -> virtual TypeInfo::cmp() 
    -> virtual final ScalarTypeInfo::cmp
      -> ScalarTypeInfoImplBase::cmp
```

In which, `ScalarTypeInfo` is final inheritance, so could be devirtualized, `ScalarTypeInfoImplBase::cmp` is actually used a function pointer and instantiated into a ScalarTypeInfo instance, also no virtual function call. So as expected there's no virtual function call in `PlainPageDecoder::seek_at_or_after_value`.

Further more, the call stack is inlined, even that some code is at `.cpp` file, I think it's the linker's contribution.

```c++
Disassembly of section .text._ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb:

0000000000000000 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb>:
55                   	push   %rbp
48 89 e5             	mov    %rsp,%rbp
41 54                	push   %r12
53                   	push   %rbx
48 81 ec a0 00 00 00 	sub    $0xa0,%rsp
48 89 bd 68 ff ff ff 	mov    %rdi,-0x98(%rbp)
48 89 b5 60 ff ff ff 	mov    %rsi,-0xa0(%rbp)
48 89 95 58 ff ff ff 	mov    %rdx,-0xa8(%rbp)
48 89 8d 50 ff ff ff 	mov    %rcx,-0xb0(%rbp)
48 8b 85 60 ff ff ff 	mov    -0xa0(%rbp),%rax
0f b6 40 28          	movzbl 0x28(%rax),%eax
83 f0 01             	xor    $0x1,%eax
0f b6 c0             	movzbl %al,%eax
bb 00 00 00 00       	mov    $0x0,%ebx
48 85 c0             	test   %rax,%rax
74 5f                	je     a4 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0xa4>
48 8d 45 80          	lea    -0x80(%rbp),%rax
ba 9b 00 00 00       	mov    $0x9b,%edx
be 00 00 00 00       	mov    $0x0,%esi
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  5b <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x5b>
bb 01 00 00 00       	mov    $0x1,%ebx
48 8d 45 80          	lea    -0x80(%rbp),%rax
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  6c <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x6c>
be 00 00 00 00       	mov    $0x0,%esi
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  79 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x79>
be 00 00 00 00       	mov    $0x0,%esi
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  86 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x86>
49 89 c4             	mov    %rax,%r12
48 8d 45 9f          	lea    -0x61(%rbp),%rax
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  95 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x95>
48 8d 45 9f          	lea    -0x61(%rbp),%rax
4c 89 e6             	mov    %r12,%rsi
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  a4 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0xa4>
84 db                	test   %bl,%bl
74 0c                	je     b4 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0xb4>
48 8d 45 80          	lea    -0x80(%rbp),%rax
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  b4 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0xb4>
48 8b 85 60 ff ff ff 	mov    -0xa0(%rbp),%rax
8b 40 2c             	mov    0x2c(%rax),%eax
85 c0                	test   %eax,%eax
75 2c                	jne    ee <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0xee>
48 8d 45 a0          	lea    -0x60(%rbp),%rax
be 00 00 00 00       	mov    $0x0,%esi
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  d3 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0xd3>
48 8b 85 68 ff ff ff 	mov    -0x98(%rbp),%rax
48 8d 55 a0          	lea    -0x60(%rbp),%rdx
48 89 d6             	mov    %rdx,%rsi
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  e9 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0xe9>
e9 e0 01 00 00       	jmpq   2ce <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x2ce>
48 c7 45 e8 00 00 00 	movq   $0x0,-0x18(%rbp)
00 
48 8b 85 60 ff ff ff 	mov    -0xa0(%rbp),%rax
8b 40 2c             	mov    0x2c(%rax),%eax
89 c0                	mov    %eax,%eax
48 89 45 e0          	mov    %rax,-0x20(%rbp)
48 c7 45 d8 00 00 00 	movq   $0x0,-0x28(%rbp)
00 
48 8d 85 70 ff ff ff 	lea    -0x90(%rbp),%rax
be 01 00 00 00       	mov    $0x1,%esi
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  122 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x122>
48 8b 45 e8          	mov    -0x18(%rbp),%rax
48 3b 45 e0          	cmp    -0x20(%rbp),%rax
0f 83 8d 00 00 00    	jae    1bd <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x1bd>
48 8b 45 e0          	mov    -0x20(%rbp),%rax
48 2b 45 e8          	sub    -0x18(%rbp),%rax
48 d1 e8             	shr    %rax
48 89 c2             	mov    %rax,%rdx
48 8b 45 e8          	mov    -0x18(%rbp),%rax
48 01 d0             	add    %rdx,%rax
48 89 45 d0          	mov    %rax,-0x30(%rbp)
48 8b 85 60 ff ff ff 	mov    -0xa0(%rbp),%rax
48 83 c0 08          	add    $0x8,%rax
48 8b 55 d0          	mov    -0x30(%rbp),%rdx
48 83 c2 04          	add    $0x4,%rdx
48 89 d6             	mov    %rdx,%rsi
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  167 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x167>
48 89 45 d8          	mov    %rax,-0x28(%rbp)
48 8d 85 70 ff ff ff 	lea    -0x90(%rbp),%rax
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  17a <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x17a>
48 8b 10             	mov    (%rax),%rdx
48 83 c2 08          	add    $0x8,%rdx
4c 8b 02             	mov    (%rdx),%r8
48 8b 95 58 ff ff ff 	mov    -0xa8(%rbp),%rdx
48 8b 4d d8          	mov    -0x28(%rbp),%rcx
48 89 ce             	mov    %rcx,%rsi
48 89 c7             	mov    %rax,%rdi
41 ff d0             	callq  *%r8
c1 e8 1f             	shr    $0x1f,%eax
84 c0                	test   %al,%al
74 11                	je     1b0 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x1b0>
48 8b 45 d0          	mov    -0x30(%rbp),%rax
48 83 c0 01          	add    $0x1,%rax
48 89 45 e8          	mov    %rax,-0x18(%rbp)
e9 72 ff ff ff       	jmpq   122 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x122>
48 8b 45 d0          	mov    -0x30(%rbp),%rax
48 89 45 e0          	mov    %rax,-0x20(%rbp)
e9 65 ff ff ff       	jmpq   122 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x122>
48 8b 85 60 ff ff ff 	mov    -0xa0(%rbp),%rax
8b 40 2c             	mov    0x2c(%rax),%eax
89 c0                	mov    %eax,%eax
48 39 45 e8          	cmp    %rax,-0x18(%rbp)
72 2c                	jb     1fb <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x1fb>
48 8d 45 b0          	lea    -0x50(%rbp),%rax
be 00 00 00 00       	mov    $0x0,%esi
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  1e0 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x1e0>
48 8b 85 68 ff ff ff 	mov    -0x98(%rbp),%rax
48 8d 55 b0          	lea    -0x50(%rbp),%rdx
48 89 d6             	mov    %rdx,%rsi
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  1f6 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x1f6>
e9 8d 00 00 00       	jmpq   288 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x288>
48 8b 85 60 ff ff ff 	mov    -0xa0(%rbp),%rax
48 83 c0 08          	add    $0x8,%rax
48 8b 55 e8          	mov    -0x18(%rbp),%rdx
48 83 c2 04          	add    $0x4,%rdx
48 89 d6             	mov    %rdx,%rsi
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  219 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x219>
48 89 45 c8          	mov    %rax,-0x38(%rbp)
48 8d 85 70 ff ff ff 	lea    -0x90(%rbp),%rax
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  22c <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x22c>
48 8b 10             	mov    (%rax),%rdx
48 83 c2 08          	add    $0x8,%rdx
4c 8b 02             	mov    (%rdx),%r8
48 8b 95 58 ff ff ff 	mov    -0xa8(%rbp),%rdx
48 8b 4d c8          	mov    -0x38(%rbp),%rcx
48 89 ce             	mov    %rcx,%rsi
48 89 c7             	mov    %rax,%rdi
41 ff d0             	callq  *%r8
85 c0                	test   %eax,%eax
0f 94 c0             	sete   %al
84 c0                	test   %al,%al
74 0c                	je     25f <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x25f>
48 8b 85 50 ff ff ff 	mov    -0xb0(%rbp),%rax
c6 00 01             	movb   $0x1,(%rax)
eb 0a                	jmp    269 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x269>
48 8b 85 50 ff ff ff 	mov    -0xb0(%rbp),%rax
c6 00 00             	movb   $0x0,(%rax)
48 8b 45 e8          	mov    -0x18(%rbp),%rax
89 c2                	mov    %eax,%edx
48 8b 85 60 ff ff ff 	mov    -0xa0(%rbp),%rax
89 50 30             	mov    %edx,0x30(%rax)
48 8b 85 68 ff ff ff 	mov    -0x98(%rbp),%rax
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  288 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x288>
48 8d 85 70 ff ff ff 	lea    -0x90(%rbp),%rax
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  297 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x297>
eb 35                	jmp    2ce <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x2ce>
84 db                	test   %bl,%bl
74 0c                	je     2a9 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x2a9>
48 8d 45 80          	lea    -0x80(%rbp),%rax
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  2a9 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x2a9>
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  2b1 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x2b1>
48 89 c3             	mov    %rax,%rbx
48 8d 85 70 ff ff ff 	lea    -0x90(%rbp),%rax
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  2c3 <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x2c3>
48 89 d8             	mov    %rbx,%rax
48 89 c7             	mov    %rax,%rdi
e8 00 00 00 00       	callq  2ce <_ZN9starrocks16PlainPageDecoderILNS_9FieldTypeE1EE22seek_at_or_after_valueEPKvPb+0x2ce>
48 8b 85 68 ff ff ff 	mov    -0x98(%rbp),%rax
48 81 c4 a0 00 00 00 	add    $0xa0,%rsp
5b                   	pop    %rbx
41 5c                	pop    %r12
5d                   	pop    %rbp
c3                   	retq   

```

Reply:
It looks like this analysis is not quite right, here it looks like the dispatch code

Reply:
Looks like there's actually a virtual function call:
```c++
4c 8b 02             	mov    (%rdx),%r8
48 8b 95 58 ff ff ff 	mov    -0xa8(%rbp),%rdx
48 8b 4d d8          	mov    -0x28(%rbp),%rcx
48 89 ce             	mov    %rcx,%rsi
48 89 c7             	mov    %rax,%rdi
41 ff d0             	callq  *%r8
```